File: wrappers/mvc/src/Kendo.Mvc/UI/ToolBar/ToolBarItem.cs
namespace Kendo.Mvc.UI
{
    using System;
    using System.Linq;
    using System.Collections.Generic;
    using System.Web.Routing;
    using Kendo.Mvc.Extensions;

    public class ToolBarItem : JsonObject
    {
        public ToolBarItem()
        {
            //>> Initialization
        
            Buttons = new List<ToolBarItemButton>();
                
            MenuButtons = new List<ToolBarItemMenuButton>();
                
        //<< Initialization
        }

        //>> Fields
        
        public List<ToolBarItemButton> Buttons
        {
            get;
            set;
        }
        
        public string Click { get; set; }
        
        public bool? Enable { get; set; }
        
        public string Group { get; set; }
        
        public string Icon { get; set; }
        
        public string Id { get; set; }
        
        public string ImageUrl { get; set; }
        
        public List<ToolBarItemMenuButton> MenuButtons
        {
            get;
            set;
        }
        
        public string OverflowTemplate { get; set; }

        public string OverflowTemplateId { get; set; }
        
        public bool? Primary { get; set; }
        
        public bool? Selectable { get; set; }
        
        public string SpriteCssClass { get; set; }
        
        public string Template { get; set; }

        public string TemplateId { get; set; }
        
        public string Text { get; set; }
        
        public bool? Togglable { get; set; }
        
        public string Toggle { get; set; }
        
        public string Url { get; set; }
        
        public CommandType? Type { get; set; }
        
        public ShowIn? ShowText { get; set; }
        
        public ShowIn? ShowIcon { get; set; }
        
        public ShowInOverflowPopup? Overflow { get; set; }
        
        //<< Fields

        protected override void Serialize(IDictionary<string, object> json)
        {
            //>> Serialization
        
            var buttons = Buttons.ToJson();
            if (buttons.Any())
            {
                json["buttons"] = buttons;
            }
                
            if (Click.HasValue())
            {
                json["click"] = Click;
            }
            
            if (Enable.HasValue)
            {
                json["enable"] = Enable;
            }
                
            if (Group.HasValue())
            {
                json["group"] = Group;
            }
            
            if (Icon.HasValue())
            {
                json["icon"] = Icon;
            }
            
            if (Id.HasValue())
            {
                json["id"] = Id;
            }
            
            if (ImageUrl.HasValue())
            {
                json["imageUrl"] = ImageUrl;
            }
            
            var menuButtons = MenuButtons.ToJson();
            if (menuButtons.Any())
            {
                json["menuButtons"] = menuButtons;
            }
                
            if (!string.IsNullOrEmpty(OverflowTemplateId))
            {
                json["overflowTemplate"] = new ClientHandlerDescriptor {
                    HandlerName = string.Format(
                        "jQuery('#{0}').html()",
                        OverflowTemplateId
                    )
                };
            }
            else if (!string.IsNullOrEmpty(OverflowTemplate))
            {
                json["overflowTemplate"] = OverflowTemplate;
            }
                
            if (Primary.HasValue)
            {
                json["primary"] = Primary;
            }
                
            if (Selectable.HasValue)
            {
                json["selectable"] = Selectable;
            }
                
            if (SpriteCssClass.HasValue())
            {
                json["spriteCssClass"] = SpriteCssClass;
            }
            
            if (!string.IsNullOrEmpty(TemplateId))
            {
                json["template"] = new ClientHandlerDescriptor {
                    HandlerName = string.Format(
                        "jQuery('#{0}').html()",
                        TemplateId
                    )
                };
            }
            else if (!string.IsNullOrEmpty(Template))
            {
                json["template"] = Template;
            }
                
            if (Text.HasValue())
            {
                json["text"] = Text;
            }
            
            if (Togglable.HasValue)
            {
                json["togglable"] = Togglable;
            }
                
            if (Toggle.HasValue())
            {
                json["toggle"] = Toggle;
            }
            
            if (Url.HasValue())
            {
                json["url"] = Url;
            }
            
            if (Type.HasValue)
            {
                json["type"] = Type;
            }
                
            if (ShowText.HasValue)
            {
                json["showText"] = ShowText;
            }
                
            if (ShowIcon.HasValue)
            {
                json["showIcon"] = ShowIcon;
            }
                
            if (Overflow.HasValue)
            {
                json["overflow"] = Overflow;
            }
                
        //<< Serialization

            if (Type.HasValue)
            {
                json["type"] = Type.ToString().ToCamelCase();
            }
        }
    }
}

File: wrappers/php/lib/Kendo/UI/ToolBarItem.php
<?php

namespace Kendo\UI;

class ToolBarItem extends \Kendo\SerializableObject {
//>> Properties

    /**
    * Specifies the HTML attributes of a ToolBar button.
    * @param  $value
    * @return \Kendo\UI\ToolBarItem
    */
    public function attributes($value) {
        return $this->setProperty('attributes', $value);
    }

    /**
    * Adds ToolBarItemButton to the ToolBarItem.
    * @param \Kendo\UI\ToolBarItemButton|array,... $value one or more ToolBarItemButton to add.
    * @return \Kendo\UI\ToolBarItem
    */
    public function addButton($value) {
        return $this->add('buttons', func_get_args());
    }

    /**
    * Sets the click option of the ToolBarItem.
    * Specifies the click event handler of the button. Applicable only for commands of type button and splitButton.
    * @param string|\Kendo\JavaScriptFunction $value Can be a JavaScript function definition or name.
    * @return \Kendo\UI\ToolBarItem
    */
    public function click($value) {
        if (is_string($value)) {
            $value = new \Kendo\JavaScriptFunction($value);
        }

        return $this->setProperty('click', $value);
    }

    /**
    * Specifies whether the control is initially enabled or disabled. Default value is "true".
    * @param boolean $value
    * @return \Kendo\UI\ToolBarItem
    */
    public function enable($value) {
        return $this->setProperty('enable', $value);
    }

    /**
    * Assigns the button to a group. Applicable only for buttons with togglable: true.
    * @param string $value
    * @return \Kendo\UI\ToolBarItem
    */
    public function group($value) {
        return $this->setProperty('group', $value);
    }

    /**
    * Sets icon for the item. The icon should be one of the existing in the Kendo UI theme sprite.
    * @param string $value
    * @return \Kendo\UI\ToolBarItem
    */
    public function icon($value) {
        return $this->setProperty('icon', $value);
    }

    /**
    * Specifies the ID of the button.
    * @param string $value
    * @return \Kendo\UI\ToolBarItem
    */
    public function id($value) {
        return $this->setProperty('id', $value);
    }

    /**
    * If set, the ToolBar will render an image with the specified URL in the button.
    * @param string $value
    * @return \Kendo\UI\ToolBarItem
    */
    public function imageUrl($value) {
        return $this->setProperty('imageUrl', $value);
    }

    /**
    * Adds ToolBarItemMenuButton to the ToolBarItem.
    * @param \Kendo\UI\ToolBarItemMenuButton|array,... $value one or more ToolBarItemMenuButton to add.
    * @return \Kendo\UI\ToolBarItem
    */
    public function addMenuButton($value) {
        return $this->add('menuButtons', func_get_args());
    }

    /**
    * Sets the overflowTemplate option of the ToolBarItem.
    * Specifies what element will be added in the command overflow popup. Applicable only for items that have a template.
    * @param string $value The id of the element which represents the kendo template.
    * @return \Kendo\UI\ToolBarItem
    */
    public function overflowTemplateId($value) {
        $value = new \Kendo\Template($value);

        return $this->setProperty('overflowTemplate', $value);
    }

    /**
    * Sets the overflowTemplate option of the ToolBarItem.
    * Specifies what element will be added in the command overflow popup. Applicable only for items that have a template.
    * @param string $value The template content.
    * @return \Kendo\UI\ToolBarItem
    */
    public function overflowTemplate($value) {
        return $this->setProperty('overflowTemplate', $value);
    }

    /**
    * Specifies whether the button is primary. Primary buttons receive different styling.
    * @param boolean $value
    * @return \Kendo\UI\ToolBarItem
    */
    public function primary($value) {
        return $this->setProperty('primary', $value);
    }

    /**
    * Specifies if the toggle button is initially selected. Applicable only for buttons with togglable: true.
    * @param boolean $value
    * @return \Kendo\UI\ToolBarItem
    */
    public function selected($value) {
        return $this->setProperty('selected', $value);
    }

    /**
    * Defines a CSS class (or multiple classes separated by spaces) which will be used for button icon.
    * @param string $value
    * @return \Kendo\UI\ToolBarItem
    */
    public function spriteCssClass($value) {
        return $this->setProperty('spriteCssClass', $value);
    }

    /**
    * Sets the template option of the ToolBarItem.
    * Specifies what element will be added in the ToolBar wrapper. Items with template does not have a type.
    * @param string $value The id of the element which represents the kendo template.
    * @return \Kendo\UI\ToolBarItem
    */
    public function templateId($value) {
        $value = new \Kendo\Template($value);

        return $this->setProperty('template', $value);
    }

    /**
    * Sets the template option of the ToolBarItem.
    * Specifies what element will be added in the ToolBar wrapper. Items with template does not have a type.
    * @param string $value The template content.
    * @return \Kendo\UI\ToolBarItem
    */
    public function template($value) {
        return $this->setProperty('template', $value);
    }

    /**
    * Sets the text of the button.
    * @param string $value
    * @return \Kendo\UI\ToolBarItem
    */
    public function text($value) {
        return $this->setProperty('text', $value);
    }

    /**
    * Specifies if the button is togglable, e.g. has a selected and unselected state.
    * @param boolean $value
    * @return \Kendo\UI\ToolBarItem
    */
    public function togglable($value) {
        return $this->setProperty('togglable', $value);
    }

    /**
    * Sets the toggle option of the ToolBarItem.
    * Specifies the toggle event handler of the button. Applicable only for commands of type button and togglable: true.
    * @param string|\Kendo\JavaScriptFunction $value Can be a JavaScript function definition or name.
    * @return \Kendo\UI\ToolBarItem
    */
    public function toggle($value) {
        if (is_string($value)) {
            $value = new \Kendo\JavaScriptFunction($value);
        }

        return $this->setProperty('toggle', $value);
    }

    /**
    * Specifies the url to navigate to.
    * @param string $value
    * @return \Kendo\UI\ToolBarItem
    */
    public function url($value) {
        return $this->setProperty('url', $value);
    }

    /**
    * Specifies the type of the item.
    * @param string $value
    * @return \Kendo\UI\ToolBarItem
    */
    public function type($value) {
        return $this->setProperty('type', $value);
    }

    /**
    * Specifies where the text will be displayed.
    * @param string $value
    * @return \Kendo\UI\ToolBarItem
    */
    public function showText($value) {
        return $this->setProperty('showText', $value);
    }

    /**
    * Specifies where the icon will be displayed.
    * @param string $value
    * @return \Kendo\UI\ToolBarItem
    */
    public function showIcon($value) {
        return $this->setProperty('showIcon', $value);
    }

    /**
    * Specifies how the button behaves when the ToolBar is resized.
    * @param string $value
    * @return \Kendo\UI\ToolBarItem
    */
    public function overflow($value) {
        return $this->setProperty('overflow', $value);
    }

//<< Properties
}

?>
